### PR TITLE
Adding numba to allow for parallel histograms

### PR DESCRIPTION
**Pull request by AnthonyLim23**
#### Summary of work

This PR adds code to allow the histograms to be calculated in parallel using numba.
 

### To test:

The scripts in profile provide a set of tests for the data. I would suggest making sure that the stats are set to a low number and to change the `get_data` in `utils` to only return the first few files.
#### Code Review

- Is the code of an acceptable quality?
- Are the unit tests small and test the class in isolation?
- Does the documentation build [ReadTheDocs](https://readthedocs.org/projects/muondatalib/builds/)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**.

### CODE REVIEW

**Comment by AnthonyLim23** [comment]
I need to update the tests and docs

File: src/MuonDataLib/numba/stats.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    """
+    Gets the max number of threads
+    :return: the max number of threads
+    """
+    return numba.get_num_threads()
+
+
+@numba.jit(float64[:](float64, float64, float64),
+           nopython=True, fastmath=True, parallel=True)
+def get_bin_edges(a_min, a_max, delta):
```
Comment (alexhroom) [inline src/MuonDataLib/numba/stats.py]: `delta` is called `width` everywhere else, can we use `width` for a consistent variable name?

File: src/MuonDataLib/numba/stats.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            time = times[k]*conversion
+            if time <= max_time and time >= min_time:
+                j_bin = int((time - min_time) // width)
+                p = periods[k]
+                det = spec[k]
+                w = weight[k]
+                result[thread, p, det, j_bin] += w
+                N[thread] += w
+
+    return np.sum(result, axis=0, dtype=np.int32), bins, np.sum(N)
```
Comment (alexhroom) [inline src/MuonDataLib/numba/stats.py]: I see it says above we need a unique matrix for `results` because of a race condition, but the [Numba docs seem to suggest at least for N we should be safe just accumulating on N...](https://numba.readthedocs.io/en/stable/user/performance-tips.html#parallel-true)

File: src/MuonDataLib/numba/stats.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    This method creates histograms from a list of data in parallel.
+    It produces a matrix of histograms for multiple periods
+    and spectra.
+    Strictly speaking these are not histograms as they
+    are not normalised to bin width.
+    This is the language used by the users and the
+    analysis code applys the normalisation.
+
+    To avoid a race condition we need to use a
+    unique matrix on each thread, then sum them afterwards.
```
Comment (alexhroom) [inline src/MuonDataLib/numba/stats.py]: in my opinion, docstrings are user-facing so shouldn't contain implementation details (this is useless information to anyone who isn't looking at the source code to modify this function). this should be a code comment in the function.

File: src/MuonDataLib/cython_ext/event_data.pyx [hunk 4]
```diff
@@ -435,28 +436,46 @@ cdef class Events:
...
+        elif N_threads > 1:
+            hist, bins, N = para_histogram(times=times,
+                                           spec=IDs,
+                                           N_spec=self.N_spec,
+                                           periods=periods,
+                                           min_time=min_time,
+                                           max_time=max_time,
+                                           width=width,
+                                           weight=weight,
+                                           conversion=1.e-3,
```
Comment (alexhroom) [inline src/MuonDataLib/cython_ext/event_data.pyx:474]: if `conversion` is always hardcoded why is it a parameter? do we plan to make this something the user can set in future?

File: src/MuonDataLib/cython_ext/event_data.pyx [hunk 5]
```diff
@@ -435,28 +436,46 @@ cdef class Events:
...
+                                           N_threads=N_threads)
+        else:
+            raise ValueError(f'The number of threads {N_threads} is not valid')
```
Comment (alexhroom) [inline src/MuonDataLib/cython_ext/event_data.pyx]: for readability:
```suggestion
            raise ValueError(f'The number of threads ({N_threads}) is not valid')
```

File: src/MuonDataLib/numba/stats.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
+import numpy as np
+
+import numba
+from numba import int32, float64, int64
+
+
+def get_max_threads():
+    """
+    Gets the max number of threads
```
Comment (alexhroom) [inline src/MuonDataLib/numba/stats.py]: the max number of threads for what? on the machine? set by the user? this docstring doesn't add anything that the function name hasn't already told me

File: src/MuonDataLib/numba/stats.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+    for thread in numba.prange(N_threads):
+        # get slices
+        start = thread*chunk
+        stop = start + chunk
+        if thread == N_threads-1:
+            stop = len(times)
+        for k in range(start, stop):
+            time = times[k]*conversion
+            if time <= max_time and time >= min_time:
```
Comment (alexhroom) [inline src/MuonDataLib/numba/stats.py:93]: wonder if it would be faster to convert the whole `times` array and filter out all times outside of the min/max time before this loop so that it gets vectorised?

File: src/MuonDataLib/cython_ext/filter.pyx [hunk 8]
```diff
@@ -146,7 +146,7 @@ cpdef good_periods(int[:] f_start, int[:] f_end, int[:] start_index, int[:] peri
     """
 
     cdef Py_ssize_t start = 0
-    cdef Py_ssize_t M, j, end, dm, filter_start
+    cdef Py_ssize_t M, j, k, end, dm, filter_start
```
Comment (alexhroom) [inline src/MuonDataLib/cython_ext/filter.pyx:149]: is `k` unused?

File: src/MuonDataLib/cython_ext/filter.pyx [hunk 9]
```diff
@@ -178,7 +178,7 @@ cpdef good_periods(int[:] f_start, int[:] f_end, int[:] start_index, int[:] peri
     dm = N_events - start_index[N_frames-1]
     good_periods[M:M + dm] = periods[len(periods)-1]
     M += dm
-    return good_periods[:M]
+    return _good_periods[:M]
```
Comment (alexhroom) [inline src/MuonDataLib/cython_ext/filter.pyx:181]: what's going on here? `_good_periods` is set at the start to zeroes and returned, but the period changes are added to `good_periods`? am I missing some cython magic?

File: src/MuonDataLib/data/muon_data.py [hunk 10]
```diff
@@ -203,13 +203,14 @@ def hist_settings_changed(self, cached_settings):
...
 
-    def histogram(self):
+    def histogram(self, N_threads=1):
         """
         A method for constructing a histogram.
         This will skip calculating the filters
         if the cache is occupied.
         If just the resolution has changed it will
         not alter the filtered values.
+        :param N_threads: the number of threads to run on
```
Comment (alexhroom) [inline src/MuonDataLib/data/muon_data.py]: can we guarantee the user knows what a 'thread' is or do we need to make explicit that this parameter is for parallelisation?

File: src/MuonDataLib/numba/stats.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+                   conversion=1.e-3,
+                   N_threads=1):
+    """
+    This method creates histograms from a list of data in parallel.
+    It produces a matrix of histograms for multiple periods
+    and spectra.
+    Strictly speaking these are not histograms as they
+    are not normalised to bin width.
+    This is the language used by the users and the
+    analysis code applys the normalisation.
```
Comment (alexhroom) [inline src/MuonDataLib/numba/stats.py]: I think these two paragraphs should be removed as all the three points they make aren't necessary:

- I have never heard (and can't find a source) that histograms are not 'strictly' histograms if they're not normalised - in fact [a non-normalised histogram is the default behaviour of `numpy.hist` which readily calls it a histogram](https://numpy.org/doc/stable/reference/generated/numpy.histogram.html). If you want to be clear I'd just suggest changing the first paragraph of this docstring to say `It produces a matrix of (non-normalised) histograms`?
- "this is the language used by the users" is only true if by "users" you mean "the Muon Group at ISIS Neutron and Muon Source" - we don't know who the "users" are in general and what they do or don't consider to be a histogram.
- "the analysis code applies the normalisation", again assumes what "the analysis code" is (I assume WiMDA) so this might not be true when we start supporting other analysis software. either:
  - all analysis code we will ever support applies the normalisation (I don't think we can ever guarantee this) in which case this comment isn't relevant as users won't care that these histograms are unnormalised, or
  - some analysis code doesn't do this, in which case the statement is not true

In short: we should just say the histogram is not normalised and assume the user knows what the consequences of that is

File: src/MuonDataLib/numba/stats.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    """
+    return numba.get_num_threads()
+
+
+@numba.jit(float64[:](float64, float64, float64),
+           nopython=True, fastmath=True, parallel=True)
+def get_bin_edges(a_min, a_max, delta):
+    """
+    A method to reproduce the result of np.arange, but in parallel.
+    I have tested this and it seems to be faster.
```
Comment (alexhroom) [inline src/MuonDataLib/numba/stats.py]: the line about your testing shouldn't be here (or should be a code comment) as this isn't relevant to the documentation. I also think this should be a clearer one-line about the actual result of the function, e.g.:

```suggestion
    Calculate histogram bin edges for a given range and width.
```

File: test/numba/numba_stats_test.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+            conversion=1e-3, N_threads=2
+        )
+
+        """
+        500ms * 1e-3 = 0.5s -> bin 0
+        1500ms * 1e-3 = 1.5s -> bin 1
+        """
+        np.testing.assert_array_equal(result[0, 0], [1, 1])
+        self.assertEqual(N, 2)
+
```
Comment (alexhroom) [inline test/numba/numba_stats_test.py:98]: Is this test really necessary? surely if `conversion` wasn't applied properly the other tests would fail and we don't need to test that Python multiplication works?

File: test/numba/numba_stats_test.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    def test_parallel(self):
+        N = get_max_threads()
+
+        log = logging.getLogger('TestStats.test_parallel')
+        log.debug("")
+        log.debug(f"{N} threads available")
+
+        self.assertEqual(get_max_threads(),
+                         numba.get_num_threads())
```
Comment (alexhroom) [inline test/numba/numba_stats_test.py:21]: useless test - what are we testing? basically just
```python
def function():
    return do_something()

assert function() == do_something()
```

and why does this test log? no other test has logging

**Review by alexhroom — CHANGES_REQUESTED** [review]
this is just an initial passthrough reading, haven't tested yet - mostly looks good functionally I just have some questions about documentation & tests :)

File: src/MuonDataLib/numba/stats.py (hunk 6 above)
Comment (alexhroom) [reply src/MuonDataLib/numba/stats.py]: i also see that this function isn't actually called anywhere...

File: src/MuonDataLib/numba/stats.py (hunk 2 above)
Comment (AnthonyLim23) [reply src/MuonDataLib/numba/stats.py]: Thats what I originally tried and I was seeing about a 0.2% drop in the counts for the histogram, but not the total counts. Whereas this method does not drop anything. All I can think of is that its not safe to accumulate the result out of order.

File: src/MuonDataLib/cython_ext/event_data.pyx (hunk 4 above)
Comment (AnthonyLim23) [reply src/MuonDataLib/cython_ext/event_data.pyx:474]: I have a vague recollection of it being something that might become a user facing parameter in the future.

File: src/MuonDataLib/data/muon_data.py (hunk 10 above)
Comment (AnthonyLim23) [reply src/MuonDataLib/data/muon_data.py]: fair point. I will update it

File: src/MuonDataLib/numba/stats.py (hunk 2 above)
Comment (alexhroom) [reply src/MuonDataLib/numba/stats.py]: to clarify: I mean just accumulating `N` out of order, but still accumulating `result` like we do right now

File: src/MuonDataLib/numba/stats.py (hunk 11 above)
Comment (AnthonyLim23) [reply src/MuonDataLib/numba/stats.py]: I believe on the defining features of a histograms is that the area of the bar as the counts. Hence, the comment. Since we dont divide by bin width, it was not a "proper" histogram as we have frequency on the y axis and not the density  https://www.bbc.co.uk/bitesize/guides/zspfcwx/revision/1

The second bullet is more a note to myself

Both Wimda and Mantid use the unnormalised histogram. In Mantid we can switch it between normalised and unnormalised after loading the data

File: test/numba/numba_stats_test.py (hunk 13 above)
Comment (AnthonyLim23) [reply test/numba/numba_stats_test.py:98]: The other tests use `conversion=1`

File: src/MuonDataLib/numba/stats.py (hunk 11 above)
Comment (alexhroom) [reply src/MuonDataLib/numba/stats.py]: [every example of a histogram on the Wikipedia page for histogram](https://en.wikipedia.org/wiki/Histogram) has frequency and not density on the y-axis...

> The second bullet is more a note to myself

Then I think it should be a code comment rather than in the docstring - users don't need to see it

File: test/numba/numba_stats_test.py (hunk 14 above)
Comment (AnthonyLim23) [reply test/numba/numba_stats_test.py:21]: This test exists as a check for us. By printing the number of threads used, we can check that the code is actually running in parallel. If `threads=1` then its not really being tested properly. I suppose we could change it to be `assert threads > 1` but that would always fail in CI if only one was available, which might be a bit heavy handed. What do you think?

File: src/MuonDataLib/numba/stats.py (hunk 6 above)
Comment (AnthonyLim23) [reply src/MuonDataLib/numba/stats.py]: This will be used when we set parallel to be the default.

File: src/MuonDataLib/numba/stats.py (hunk 7 above)
Comment (AnthonyLim23) [reply src/MuonDataLib/numba/stats.py:93]: I think I tried that with cython and it was slower. I can check it for numba.

File: src/MuonDataLib/cython_ext/filter.pyx (hunk 8 above)
Comment (AnthonyLim23) [reply src/MuonDataLib/cython_ext/filter.pyx:149]: its in one of the loops. Cython will guess the type if you dont define it, but it slows down the code.

File: src/MuonDataLib/cython_ext/filter.pyx (hunk 9 above)
Comment (AnthonyLim23) [reply src/MuonDataLib/cython_ext/filter.pyx:181]: yes, this is Cython magic. We define a numpy array `_good_periods` then make the memory view `good_periods` which is initialised to be equal the numpy array. Hence, the similar names. Since its faster to work with memory views we edit `good_periods`, but the numpy version is also updated. The change exists so we can return the numpy array instead of the memory view. 

https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html

File: src/MuonDataLib/cython_ext/filter.pyx (hunk 8 above)
Comment (AnthonyLim23) [reply src/MuonDataLib/cython_ext/filter.pyx:149]: https://github.com/ISISMuon/MuonDataLib/pull/122/changes#diff-72488d29f63cc0f74aac587f7aa16e5e81be88889de614b2911356f7d9f28137R163

File: test/numba/numba_stats_test.py (hunk 13 above)
Comment (alexhroom) [reply test/numba/numba_stats_test.py:98]: ahh, i see

File: test/numba/numba_stats_test.py (hunk 14 above)
Comment (alexhroom) [reply test/numba/numba_stats_test.py:21]: how about something like `pytest`'s [`skipif`](https://docs.pytest.org/en/stable/how-to/skipping.html#id1) or [`xfail`](https://docs.pytest.org/en/stable/how-to/skipping.html#id1) to handle when only one thread is available?

Comment (AnthonyLim23) [reply test/numba/numba_stats_test.py:21]: This is a question of what we think is the best way to handle the edge case of `N_threads=1`. We have the following options:

1. Ignore the number of threads and assume if it works in serial it works in parallel
2. Fail if it runs in serial
3. Skip if it would run in serial
4. Log the number of threads and run anyway

I went with option 4, because I thought 3 was too risky (there is a chance that we stop testing the code). Then option 2 felt too heavy handed and could slow down development. Then option 1 doesn't really test the point of the code. 

So I went with logging so we know how many threads we are using and can make informed decisions from it. However, the test always runs.

File: src/MuonDataLib/numba/stats.py (hunk 6 above)
Comment (alexhroom) [reply src/MuonDataLib/numba/stats.py]: could you change the function name or docstring to be more clear: this gets the number of threads that Numba will use (as opposed to e.g. 'max threads' being the maximum number of threads available on the machine)

**Review by alexhroom — CHANGES_REQUESTED** [review]

**Review by alexhroom — APPROVED** [review]
looks good! :)
